### ISSUE TITLE
base_dir and out of source build

### ISSUE THREAD

**Issue by aallrd** [ccache/ccache#600]
Hello,

I have been struggling with the `base_dir` feature and sharing the compiler cache across our build projects.
I think this might be because we are doing out of source builds and I would like to get your opinion on that.

I have several projects sharing a lot of common code, which is why I am trying to set up the compiler cache between them. My build servers are setup like that:

```
# some user is working on project00 and has synced its code on the server
/SRC/project00/...
/BUILD/project00/...

# another user is working on project01 and has synced its code on the server
/SRC/project01/...
/BUILD/project01/...
```

Our build system for each project is using `/BUILD/projectXX/...`  as working directory when calling `g++`, referencing sources, headers,... as absolute paths from `/SRC/projectXX/...`.

So when setting `base_dir = /SRC`, the absolute paths on the command line get translated to something like `../../SRC/projectXX/...` as per the documentation:

> If an absolute path matches the prefix it is rewritten as relative to the current working directory, not to base_dir.

```
CCACHE_BASE_DIR=/SRC
CWD=/BUILD/project00
[2020-06-02T12:17:59.179055 13454] Running real compiler
[2020-06-02T12:17:59.179092 13454] Executing /usr/bin/g++ -I../../SRC/project00/include -c ../../SRC/project00/example.cpp -o /BUILD/project00/example.o
```
But since we still have `projectXX` in the path, I assume that the computed hash will not allow the cached data to be re-used between projects.

I tried to rename `projectXX` to a known prefix `project` and the compiler cache was working. This however prevents me from having several projects on the same server.

Do you know if this is a known issue of this kind of setup and if there is a way I could work around that?

Thank you for your help.

**Comment by jrosdahl** [comment]
I don't see a good way to solve that for "direct hits" with the given setup. You should be able to get "preprocessed hits", though.

Commit e8444a3893cb339d97e25fd70496e448b3da18bb on master (not part of a released ccache version yet, see also #491) makes it possible to put a symlink to the source directory inside the build directory (e.g. `/BUILD/project00/src -> /SRC/project00`), but then the build system must of course use `/BUILD/project00/src` instead of `/SRC/project00` for the source code, or not use absolute paths at all.

**Comment by aallrd** [comment]
Thank you for your reply.
That's what I feared, the current setup prevents me from benefiting fully from ccache.
Yet another reason to change the way we build our projects 😀.
